src/app/CASESessionManager.h

```diff
@@ -0,0 +1,120 @@
+/*
+ *
+ *    Copyright (c) 2020-2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/OperationalDeviceProxy.h>
+#include <lib/core/CHIPConfig.h>
+#include <lib/core/CHIPCore.h>
+#include <lib/dnssd/DnssdCache.h>
+#include <lib/support/Pool.h>
+#include <messaging/ExchangeMgrDelegate.h>
+
+#include <lib/dnssd/Resolver.h>
+
+namespace chip {
+
+struct CASESessionManagerConfig
+{
+    DeviceProxyInitParams sessionInitParams;
+    Dnssd::DnssdCache<CHIP_CONFIG_MDNS_CACHE_SIZE> * dnsCache = nullptr;
+};
+
+/**
+ * This class provides the following
+ * 1. Manage a pool of operational device proxy objects for peer nodes that have active message exchange with the local node.
+ * 2. The pool contains atmost one device proxy object for a given peer node.
+ * 3. API to lookup an existing proxy object, or allocate a new one by triggering session establishment with the peer node.
+ * 4. During session establishment, trigger node ID resolution (if needed), and update the DNS-SD cache (if resolution is
+ * successful)
+ */
+class CASESessionManager : public Messaging::ExchangeMgrDelegate, public Dnssd::ResolverDelegate
+{
+public:
+    CASESessionManager(CASESessionManagerConfig & params)
+    {
+        VerifyOrReturn(params.sessionInitParams.Validate() == CHIP_NO_ERROR);
+
+        mConfig      = params;
+        mInitialized = true;
+    }
+
+    virtual ~CASESessionManager() {}
+
+    void Shutdown() { mInitialized = false; }
```

Comment:
What do we gain with this Shutdown method, given there is no symmetric Init?  It seems like this class would be fine with just its constructor and destructor.